core/executor/src/precompiles/modexp.rs

```diff
@@ -64,17 +63,30 @@ impl PrecompileContract for ModExp {
     }
 
     fn gas_cost(input: &[u8]) -> u64 {
-        match LargeNumber::parse(input) {
-            Ok(large_number) => {
-                let dynamic_gas =
-                    large_number.multiplication_complexity() * large_number.iterator_count() / 3u64;
-
-                dynamic_gas
-                    .max(Integer::from(Self::MIN_GAS))
-                    .saturating_as()
-            }
-            Err(_) => u64::MAX,
-        }
+        let calc = || -> Result<u64, PrecompileFailure> {
+            let base_size = get_data(input, 0, 32)?.saturating_as::<usize>();
+            let exponent_size = get_data(input, 32, 32)?.saturating_as::<usize>();
+            let modulo_size = get_data(input, 64, 32)?.saturating_as::<usize>();
+            let data = if input.len() > 96 {
+                &input[96..]
+            } else {
+                &input[0..0]
+            };
+
+            // use 32 as exp fixed size
+            let exponent = get_data(data, base_size, 32)?;
```

Comment:
Are we sure about this? What if `exponent_size < 32`?

Reply:
it will be padded with zero

Reply:
Shouldn't it be reading at most 32 bytes?

Reply:
The current implementation can only read 32 bytes. If it is less than 32, fill it with zero

Reply:
But when `exponent_size < 32`, we shouldn't right pad with zero, that'll make the exponent larger than the actual value.

Reply:
https://github.com/ethereum/go-ethereum/blob/a03490c6b2ff0e1d9a1274afdbe087a695d533eb/core/vm/contracts.go#L306-L314

it should fill with zero

Reply:
OK, let me rephrase. It's not about padding, it's about how long the exponent variable should be. In the linked code, when `expLen <= 32`, only `expLen` bytes are read.